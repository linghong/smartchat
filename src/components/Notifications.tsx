import React, { FC } from 'react'

import { InputErrors, UploadErrors } from '@/src/types/common'
import Notification from '@/src/components/Notification'

interface Notifications {
  isLoading: boolean
  loadingMessage: string
  successMessage: string | null
  errorMessage: string | null
  uploadErrors?: UploadErrors
  inputErrors?: InputErrors
}

const Notifications: FC<Notifications> = ({
  isLoading,
  loadingMessage = 'Uploading your data...',
  successMessage,
  errorMessage,
  uploadErrors,
  inputErrors,
}) => {
  const notifications = []

  if (isLoading) {
    notifications.push(
      <Notification key="loading" type="loading" message={loadingMessage} />,
    )
  }

  if (successMessage) {
    notifications.push(
      <Notification key="success" type="success" message={successMessage} />,
    )
  }

  if (errorMessage) {
    notifications.push(
      <Notification key="error" type="error" message={errorMessage} />,
    )
  }

  if (uploadErrors)
    Object.keys(uploadErrors).forEach(key => {
      notifications.push(
        <Notification
          key={`uploaderror-${key}`}
          type="error"
          message={uploadErrors[key]}
        />,
      )
    })

  if (inputErrors)
    Object.keys(inputErrors).forEach(key => {
      notifications.push(
        <Notification
          key={`inputerror-${key}`}
          type="error"
          message={inputErrors[key]}
        />,
      )
    })

  return notifications
}

export default Notifications
