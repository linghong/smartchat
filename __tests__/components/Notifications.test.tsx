import React from 'react';
import { render, screen } from '@testing-library/react';
import Notifications from '@/src/components/Notifications';

describe('Notifications Component', () => {
  it('renders loading notification when isLoading is true', () => {
    render(
      <Notifications
        isLoading={true}
        loadingMessage=""
        successMessage={null}
        errorMessage={null}
      />
    );
    expect(screen.getByRole('status')).toBeInTheDocument();
  });

  it('renders loading notification with custom message when provided', () => {
    const { getByText } = render(
      <Notifications
        isLoading={true}
        loadingMessage="Custom loading message"
        successMessage={null}
        errorMessage={null}
      />
    );
    expect(getByText('Custom loading message')).toBeInTheDocument();
  });

  it('renders success message when provided', () => {
    const { getByText } = render(
      <Notifications
        isLoading={false}
        loadingMessage=""
        successMessage="Success"
        errorMessage={null}
      />
    );
    expect(getByText('Success')).toBeInTheDocument();
  });

  it('renders error message when provided', () => {
    const { getByText } = render(
      <Notifications
        isLoading={false}
        loadingMessage=""
        successMessage={null}
        errorMessage="Error"
      />
    );
    expect(getByText('Error')).toBeInTheDocument();
  });

  it('renders upload errors when provided', () => {
    const uploadErrors = { file: 'File is too large' };
    const { getByText } = render(
      <Notifications
        isLoading={false}
        loadingMessage=""
        successMessage={null}
        errorMessage={null}
        uploadErrors={uploadErrors}
      />
    );
    expect(getByText('File is too large')).toBeInTheDocument();
  });

  it('renders input errors when provided', () => {
    const inputErrors = { name: 'Name is required' };
    const { getByText } = render(
      <Notifications
        isLoading={false}
        loadingMessage=""
        successMessage={null}
        errorMessage={null}
        inputErrors={inputErrors}
      />
    );
    expect(getByText('Name is required')).toBeInTheDocument();
  });

  it('renders multiple notifications when multiple props are provided', () => {
    const uploadErrors = { file: 'File is too large' };
    const inputErrors = { name: 'Name is required' };
    const { getByText } = render(
      <Notifications
        isLoading={true}
        loadingMessage="Loading..."
        successMessage="Success"
        errorMessage="Error"
        uploadErrors={uploadErrors}
        inputErrors={inputErrors}
      />
    );
    expect(getByText('Loading...')).toBeInTheDocument();
    expect(getByText('Success')).toBeInTheDocument();
    expect(getByText('Error')).toBeInTheDocument();
    expect(getByText('File is too large')).toBeInTheDocument();
    expect(getByText('Name is required')).toBeInTheDocument();
  });

  it('matches snapshot with multiple notifications', () => {
    const uploadErrors = { file: 'File is too large' };
    const inputErrors = { name: 'Name is required' };
    const { asFragment } = render(
      <Notifications
        isLoading={true}
        loadingMessage="Loading..."
        successMessage="Success"
        errorMessage="Error"
        uploadErrors={uploadErrors}
        inputErrors={inputErrors}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
