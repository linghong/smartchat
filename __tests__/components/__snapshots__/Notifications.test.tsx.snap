// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications Component matches snapshot with multiple notifications 1`] = `
<DocumentFragment>
  <div
    class="p-4"
  >
    <p
      aria-live="assertive"
      class="bold text-gray-600"
      role="status"
    >
      Loading...
    </p>
  </div>
  <div
    class="p-4"
  >
    <p
      aria-live="polite"
      class="bold text-green-600"
      role="status"
    >
      Success
    </p>
  </div>
  <div
    class="p-4"
  >
    <p
      aria-live="assertive"
      class="bold text-red-600"
      role="alert"
    >
      Error
    </p>
  </div>
  <div
    class="p-4"
  >
    <p
      aria-live="assertive"
      class="bold text-red-600"
      role="alert"
    >
      File is too large
    </p>
  </div>
  <div
    class="p-4"
  >
    <p
      aria-live="assertive"
      class="bold text-red-600"
      role="alert"
    >
      Name is required
    </p>
  </div>
</DocumentFragment>
`;
